Introducing "STOPPED" replicas list into Helix InstanceConfig

jsjtzyy opened the pull request:
-Introduce "STOPPED" list field in InstanceConfig for Helix.
-Adding setReplicaStoppedState method in Delegate and HelixParticipant.
-Allow BlobStoreControlRequest to update replica stopped state in Helix.
-Rename WriteStatusDelegate to ReplicaStatusDelegate.
-Verified in Perf cluster

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplica.java:
```diff
@@ -95,14 +96,18 @@ void setSealedState(boolean isSealed) {
     this.isSealed = isSealed;
   }
 
+  void setStoppedState(boolean isStopped) {
```
This method is not used currently but might be helpful in the future PR when we would like dynamically change replica state on datanode.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/clustermap/ClusterParticipant.java:
```diff
@@ -38,6 +38,13 @@
+  /**
+   * Set or reset the stopped state of the given replica.
+   * @param replicaId the {@link ReplicaId}
+   * @param isStopped if true, the replica will be marked as stopped; otherwise it will be marked as started.
```
`@return`?

jsjtzyy replied:
added

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/Replica.java:
```diff
@@ -32,6 +32,7 @@
 class Replica implements ReplicaId {
   private final Partition partition;
   private Disk disk;
+  private volatile boolean isStopped = false;
```
there is probably no point having this here unless there is a setter method, right?

jsjtzyy replied:
A setter method is added for future use.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixParticipantTest.java:
```diff
@@ -160,6 +167,95 @@ public void testSetReplicaSealedState() throws IOException {
+    try {
+      helixParticipant.setReplicaStoppedState(nonAmbryReplica, true);
+      fail("Expected an IllegalArgumentException here");
+    } catch (Exception e) {
```
`catch (IllegalArgumentException e)` (don't need the `assertTrue()`)

jsjtzyy replied:
fixed

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixParticipantTest.java:
```diff
@@ -160,6 +167,95 @@ public void testSetReplicaSealedState() throws IOException {
+    try {
+      helixParticipantDummy.setReplicaStoppedState(replicaId1, true);
+      fail("Expected an IllegalStateException here");
+    } catch (Exception e) {
```
`catch (IllegalStateException e)`

jsjtzyy replied:
fixed

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixParticipantTest.java:
```diff
@@ -78,6 +79,12 @@ public HelixParticipantTest() throws Exception {
     props.setProperty("clustermap.cluster.name", clusterName);
     props.setProperty("clustermap.datacenter.name", "DC0");
     props.setProperty("clustermap.dcs.zk.connect.strings", zkJson.toString(2));
+    propsDummy = new Properties(props);
```
since this is only used in one test, should this be created here or in that one test? I prefer the latter

jsjtzyy replied:
good point. Moved it to specific test.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -335,10 +335,24 @@ private void checkCapacityAndUpdateWriteStatusDelegate(long totalCapacity, long
+      logger.info("Setting replica stopped state via ReplicaStatusDelegate on replica {}", replicaId);
+      return replicaStatusDelegate.setReplicaStoppedState(replicaId, isStopped);
+    }
+    logger.error("The ReplicaStatusDelegate is not instantiated");
```
i prefer a single return path - i find it easier to read and understand
```
    boolean updated = false;
    if (replicaStatusDelegate != null) {
      logger.info("Setting replica stopped state via ReplicaStatusDelegate on replica {}", replicaId);
      updated = replicaStatusDelegate.setReplicaStoppedState(replicaId, isStopped);
    } else {
      logger.error("The ReplicaStatusDelegate is not instantiated");
    }
    return updated;
```

jsjtzyy replied:
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -335,10 +335,24 @@ private void checkCapacityAndUpdateWriteStatusDelegate(long totalCapacity, long
+   * @param isStopped whether the store is stopped ({@code true}) or started.
+   * @return {@code true} if StoppedReplicas list has been updated successfully.
+   */
+  public boolean updateStoppedReplicasList(boolean isStopped) {
```
is there a reason to have a special function for this as opposed to doing it on `start()` and `shutdown()`?

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/StorageManagerTest.java:
```diff
@@ -516,5 +573,39 @@ private void generateConfigs(boolean segmentedLog) {
+  /**
+   * An extension of {@link ReplicaStatusDelegate} to help with tests.
+   */
+  private static class MockReplicaStatusDelegate extends ReplicaStatusDelegate {
```
you should be able to mock this through `Mockito` instead of writing this full blown mock. Were there any problems with that?

jsjtzyy replied:
yeah. I will take your suggestion. Thanks.

vgkholla replied:
Hmm - may still be able to use Mockito. Let's chat offline once all other comments are addressed

jsjtzyy commented on ambry-api/src/main/java/com.github.ambry/clustermap/ReplicaStatusDelegate.java:
```diff
@@ -40,4 +43,14 @@ public boolean seal(ReplicaId replicaId) {
+   * @param isStopped whether to mark the store as stopped ({@code true}).
+   * @return {@code true} if state is successful set. {@code false} if not.
+   */
+  public boolean setReplicaStoppedState(ReplicaId replicaId, boolean isStopped) {
```
sure, I will then choose to have two methods here (as seal and unseal separately)

jsjtzyy commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -335,10 +335,24 @@ private void checkCapacityAndUpdateWriteStatusDelegate(long totalCapacity, long
+   * @param isStopped whether the store is stopped ({@code true}) or started.
+   * @return {@code true} if StoppedReplicas list has been updated successfully.
+   */
+  public boolean updateStoppedReplicasList(boolean isStopped) {
```
agree. Thanks for suggestion

vgkholla commented on ambry-api/src/main/java/com.github.ambry/clustermap/ReplicaStatusDelegate.java:
```diff
@@ -0,0 +1,70 @@
+   * @param replicaIds a list of replicas whose status would be set to started.
+   * @return {@code true} if replica is successfully unmarked and becomes started. {@code false} if not.
+   */
+  public boolean unmarkStopped(List<ReplicaId> replicaIds) {
```
`markStarted()`?

vgkholla replied:
I would actually prefer `markDown()` and `markUp()` - @pnarayanan, thoughts?

jsjtzyy replied:
Sounds good. Let us discuss it with @pnarayanan and reach an agreement before making changes.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -136,6 +138,40 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
+        throw new IllegalArgumentException(
+            "HelixParticipant only works with the AmbryReplica implementation of ReplicaId");
+      }
+      replicasToUpdate.add(replicaId.getPartitionId().toPathString());
```
should this be called `partitionsToUpdate`?

jsjtzyy replied:
actually, i think `replicasToUpdate` might be slightly better because the `stopped` list is used for single datanode not for cluster-wide use. Hence, we are more interested in the replicas residing on this datanode and don't care about other replicas of this specific partition. Does it sound good for you?

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -136,6 +138,40 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
+      Set<String> stoppedReplicas = new HashSet<>(stoppedList);
+      if (!isStopped) {
+        replicasToUpdate.retainAll(stoppedReplicas);
+        logger.info("Removing replicas {} from stopped list", Arrays.toString(replicasToUpdate.toArray()));
```
I think it might be useful to check whether `replicasToUpdate` is non zero size here so as to avoid "updates" that don't change anything.

vgkholla replied:
Also need a test that verifies that

vgkholla replied:
Same for the other way round i.e stopping (don't do the operation if nothing changes)

jsjtzyy replied:
yeah. good point! I will take your advice and make changes as well as add unit tests.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -136,6 +138,40 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
+      if (stoppedList == null) {
+        stoppedList = new ArrayList<>();
+      }
+      Set<String> stoppedReplicas = new HashSet<>(stoppedList);
```
let's rename these variables - the names are too close and confusing

jsjtzyy replied:
yes, they are a little confusing. I will change them to avoid misunderstanding.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/StaticClusterAgentsFactory.java:
```diff
@@ -88,6 +89,21 @@ public void close() {
+
+        @Override
+        public List<String> getSealedReplicas() {
+          return new ArrayList<>();
```
minor: you can return `Collections.emptyList()` instead

jsjtzyy replied:
fixed. Thanks!

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -335,10 +336,28 @@ private void checkCapacityAndUpdateWriteStatusDelegate(long totalCapacity, long
+   * @param isStopped whether the store is stopped ({@code true}) or started.
+   * @return {@code true} if StoppedReplicas list has been updated successfully.
+   */
+  boolean updateStoppedReplicasList(boolean isStopped) {
```
if we are still stopping one store at a time, wouldn't it defeat the purpose of having a batch API? How do we leverage the batch API?

cc: @pnarayanan

jsjtzyy replied:
Since this method is in `BlobStore.java`, I think it naturally serves a single blobstore. If we need a batch process, we can directly invoke `ReplicaStatusDelegate.markStopped(List<ReplicaId> replicaIds)` because both `StorageManager` and `DiskManager` can access methods in`ReplicaStatusDelegate`.

pnarayanan replied:
The last time @jsjtzyy and I discussed this, a potential use case was marking all stores on a disk as stopped, when there is a disk failure. Like @jsjtzyy says, this will require the `DiskManager` or `StorageManager` invoking the api directly.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -136,6 +138,40 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
+        setResult = setStoppedReplicas(stoppedList);
+      } else {
+        stoppedReplicas.retainAll(replicasToUpdate);
+        replicasToUpdate.removeAll(stoppedReplicas);
```
What you want is UNION (for markStop) and SUBTRACTION (for unmarkStop). For sets, these are `addAll()` and `removeAll()`, respectively. So I think this would suffice:
```java
if (markStopped) { // add to STOPPED list:
  stoppedSet.addAll(replicasToUpdateSet); // union is basically addAll() for sets.
  setResult = setStoppedReplicas(Arrays.asList(stoppedSet));
} else { // remove from STOPPED list:
  stoppedList.removeAll(replicasToUpdateSet);
  setResult = setStoppedReplicas(stoppedList);
}
```

jsjtzyy replied:
Thanks for pointing it out. Actually, I am making changes on this piece code and might use the `replicasToUpdate` after `removeAll()` to determine if we need to invoke `setStoppedReplicas`. When `replicasToUpdate.isEmpty()`, it's unnecessary to update the instanceConfig in Helix again. (I will find a chance to discuss it with you offline)

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -147,6 +183,36 @@ public void close() {
   }
 
+  /**
+   * Get the list of sealed replicas from the HelixAdmin. This method is called only after the helixAdministrationLock
```
The statement "this method is only called after the helixAdministrationLock is taken" is not true now that this is a public method. Perhaps that statement is unnecessary.

vgkholla replied:
@pnarayanan @jsjtzyy There was never a "get" before the current change right? (I don't think the get needs a lock, does it?) The put still needs to be under a lock - maybe this is just a copy-paste error?

jsjtzyy replied:
@vgkholla yeah, my bad. will fix it.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -147,6 +183,36 @@ public void close() {
+  }
+
+  /**
+   * Get the list of stopped replicas from the HelixAdmin. This method is called only after the helixAdministrationLock
```
Same as above.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/Replica.java:
```diff
@@ -32,6 +32,7 @@
 class Replica implements ReplicaId {
   private final Partition partition;
   private Disk disk;
+  private volatile boolean isStopped = false;
```
are you sure this variable needs to be `volatile`?  I don't see any ordering issues.

jsjtzyy replied:
Not quite sure but it's safe to make it volatile to avoid issues on visibility of variable in current memory model.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -574,7 +593,7 @@ public void shutdown() throws StoreException {
    */
   DiskSpaceRequirements getDiskSpaceRequirements() throws StoreException {
-    checkStarted();
+    //checkStarted();
```
?

jsjtzyy replied:
fixed

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/DiskManager.java:
```diff
@@ -80,11 +82,12 @@
     diskSpaceAllocator = new DiskSpaceAllocator(diskManagerConfig.diskManagerEnableSegmentPooling,
         new File(disk.getMountPath(), diskManagerConfig.diskManagerReserveFileDirName),
         diskManagerConfig.diskManagerRequiredSwapSegmentsPerSize, metrics);
+    this.stoppedReplicas = stoppedReplicas == null ? new ArrayList<>() : stoppedReplicas;
```
perhaps `Collections.emptyList()` can be used here as well.

jsjtzyy replied:
fixed.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/DiskManager.java:
```diff
@@ -282,6 +296,22 @@ boolean shutdownBlobStore(PartitionId id) {
+      logger.error("store is not found on this disk when trying to update stoppedReplicas list");
+      return false;
+    }
+    return store.updateStoppedReplicasList(isStopped);
```
to follow up on @vgkholla's earlier comment, why relay this via the `BlobStore`? Can't this directly call into the `ReplicaStatusDelegate`?

jsjtzyy replied:
The original purpose of introducing "STOPPED" list is to mark specific BlobStore as `stopped` when we manually "stop" the BlobStore. So the update operation is performed at store level. Note that the `markStopped` operation is initiated in `AmbryRequests` and we don't want to expose blobstore to `AmbryRequests`, that's why we chose to relay operation through StorageManager and DiskManager. Also, we currently don't have `ReplicaStatusDelegate` in both `StorageManager` and `DiskManager`, which disallows us to directly call methods in `ReplicaStatusDelegate`.  (Maybe we can add `ReplicaStatusDelegate` as a field in StorageManager/DiskManager later)

jsjtzyy commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixParticipantTest.java:
```diff
@@ -191,8 +191,10 @@ public void testGetAndSetReplicaStoppedState() throws IOException, JSONException
         new HelixParticipant(new ClusterMapConfig(new VerifiableProperties(props)), helixManagerFactory);
     HelixParticipant helixParticipantDummy =
         new HelixParticipant(new ClusterMapConfig(new VerifiableProperties(propsDummy)), helixManagerFactory);
+    HelixParticipant helixParticipantSpy = Mockito.spy(helixParticipant);
```
Here we use Mockito spy() to check if the method is invoked or not in an object.

jsjtzyy commented:
@vgkholla @pnarayanan addressed the comments. In latest commit, `setBlobStoreStoppedState` will return a list of stores which fail to be updated.

I plan to address issue #941  in a separate PR. 

./gradlew build && ./gradlew test

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -136,6 +138,50 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
+        stoppedListInHelix = new ArrayList<>();
+      }
+      Set<String> stoppedListInHelixCopy = new HashSet<>(stoppedListInHelix);
+      if (!isStopped) {
```
Firstly, You should call the set `stoppedSet` rather than `stoppedListCopy`.

`removeAll()`, `retainAll()`, `addAll()`, etc. return false if the operation did not change the underlying collection, so you could use that to make this more concise:

```java
stoppedSet = new HashSet<>(stoppedListInHelix);
boolean stoppedSetUpdated = stop ? stoppedSet.addAll(replicasToUpdate) : stoppedSet.removeAll(replicasToUpdate);
return stoppedSetUpdated ? setStoppedReplicas(new ArrayList<>(stoppedSet)) : true;
```

The logging may add a few more lines though. While at it, I don't think we should log these at INFO. This is just the action in response to a request and the public access logging should be sufficient.

jsjtzyy replied:
Done. Thanks for the suggestion.

pnarayanan commented on ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java:
```diff
@@ -806,6 +807,12 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
+                List<PartitionId> failToUpdateList =
+                    storageManager.setBlobStoreStoppedState(Arrays.asList(partitionId), false);
+                if (!failToUpdateList.isEmpty()) {
+                  logger.info("Fail to remove BlobStore(s) {} from stopped list after start operation completed",
```
should this be logged at WARN level?

jsjtzyy replied:
fixed. thanks for reminding me of that.

pnarayanan commented on ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java:
```diff
@@ -843,6 +850,12 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
+                    List<PartitionId> failToUpdateList =
+                        storageManager.setBlobStoreStoppedState(Arrays.asList(partitionId), true);
+                    if (!failToUpdateList.isEmpty()) {
+                      logger.info("Fail to add BlobStore(s) {} to stopped list after stop operation completed",
```
same here. WARN level if this is unexpected.

jsjtzyy replied:
fixed

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/DiskManager.java:
```diff
@@ -80,13 +88,16 @@
         new File(disk.getMountPath(), diskManagerConfig.diskManagerReserveFileDirName),
         diskManagerConfig.diskManagerRequiredSwapSegmentsPerSize, metrics);
+    this.replicaStatusDelegate = replicaStatusDelegate;
+    this.stoppedReplicas = stoppedReplicas == null ? Collections.emptyList() : stoppedReplicas;
```
I think the caller never passes null for `stoppedReplicas` from reading the code. In general, it is best to avoid the same check when the data structures are passed around and instead do it at one place (highest level).

Perhaps a better alternative is to ensure that `ClusterMap.getStoppedReplicas()` - or whichever is the root method from where you get this list - never returns null. Makes for better readability.

jsjtzyy replied:
Changes are made in `ClusterMapUtil.getStoppedReplicas()` to ensure it never returns null.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/DiskManager.java:
```diff
@@ -104,6 +115,11 @@ void start() throws InterruptedException {
       List<Thread> startupThreads = new ArrayList<>();
       for (final Map.Entry<PartitionId, BlobStore> partitionAndStore : stores.entrySet()) {
+        if (stoppedReplicas.contains(partitionAndStore.getKey().toPathString())) {
+          // if the store is in stopped replica list, skip start process and leave it in stopped state.
```
this comment is unnecessary.

jsjtzyy replied:
removed

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/DiskManager.java:
```diff
@@ -166,6 +182,10 @@ void shutdown() throws InterruptedException {
       List<Thread> shutdownThreads = new ArrayList<>();
       for (final Map.Entry<PartitionId, BlobStore> partitionAndStore : stores.entrySet()) {
+        if (stoppedReplicas.contains(partitionAndStore.getKey().toPathString())) {
+          // if the store is in stopped replica list, skip stop process and leave it in stopped state.
```
I think the code makes this clear, comment is not required.

jsjtzyy replied:
removed

vgkholla commented on ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java:
```diff
@@ -806,6 +807,12 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
                 error = ServerErrorCode.No_Error;
                 logger.info("Store is successfully started and functional for partition: {}", partitionId);
+                List<PartitionId> failToUpdateList =
+                    storageManager.setBlobStoreStoppedState(Arrays.asList(partitionId), false);
```
minor: Can you use `Collections.singletonList()` here instead of `Arrays.asList()`?

jsjtzyy replied:
done. Thanks for suggestion.

vgkholla commented on ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java:
```diff
@@ -806,6 +807,12 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
+                List<PartitionId> failToUpdateList =
+                    storageManager.setBlobStoreStoppedState(Arrays.asList(partitionId), false);
+                if (!failToUpdateList.isEmpty()) {
+                  logger.info("Fail to remove BlobStore(s) {} from stopped list after start operation completed",
```
I don't think you need to do `Arrays.toString(failToUpdateList.toArray())`. Just adding the array as `failToUpdateList` is enough the logger should be able to resolve.

Also should this be `info` or `warn`? Seems like `warn` to me.

vgkholla replied:
Same down below

jsjtzyy replied:
done. Removed unnecessary `Arrays.toString`

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/DiskManager.java:
```diff
@@ -104,6 +115,11 @@ void start() throws InterruptedException {
       List<Thread> startupThreads = new ArrayList<>();
       for (final Map.Entry<PartitionId, BlobStore> partitionAndStore : stores.entrySet()) {
+        if (stoppedReplicas.contains(partitionAndStore.getKey().toPathString())) {
+          // if the store is in stopped replica list, skip start process and leave it in stopped state.
```
comment not required

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/DiskManager.java:
```diff
@@ -166,6 +182,10 @@ void shutdown() throws InterruptedException {
       final AtomicInteger numFailures = new AtomicInteger(0);
       List<Thread> shutdownThreads = new ArrayList<>();
       for (final Map.Entry<PartitionId, BlobStore> partitionAndStore : stores.entrySet()) {
+        if (stoppedReplicas.contains(partitionAndStore.getKey().toPathString())) {
```
Instead of this check (which may differ if the blob store was started but failed to be removed from the stopped list), isn't it just better to do
```
if (!partitionAndStore.getValue().isStarted()) {
  continue;
}
````
?

jsjtzyy replied:
yes, good point. Will take your advice.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/DiskManager.java:
```diff
@@ -282,6 +304,41 @@ boolean shutdownBlobStore(PartitionId id) {
+      updated = isStopped ? replicaStatusDelegate.markStopped(replicasToUpdate)
+          : replicaStatusDelegate.unmarkStopped(replicasToUpdate);
+    } else {
+      logger.error("The ReplicaStatusDelegate is not instantiated");
```
should this be an error (or warn)? Seems like warn to me

jsjtzyy replied:
done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/StorageManager.java:
```diff
@@ -59,11 +59,13 @@
     metrics = new StorageManagerMetrics(registry);
     StoreMetrics storeMainMetrics = new StoreMetrics(registry);
     StoreMetrics storeUnderCompactionMetrics = new StoreMetrics("UnderCompaction", registry);
+    List<String> stoppedReplicas =
```
use `Collections.emptyList()`?

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -136,6 +138,50 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
   }
 
+  @Override
+  public boolean setReplicaStoppedState(List<ReplicaId> replicaIds, boolean isStopped) {
```
fixed

jsjtzyy commented:
@vgkholla @pnarayanan  addressed the comments. Thanks for your suggestions. 
./gradlew build && ./gradlew test  (passed)

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -136,6 +138,36 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
+    synchronized (helixAdministrationLock) {
+      logger.trace("Getting stopped replicas from instanceConfig");
+      List<String> stoppedListInHelix = getStoppedReplicas();
+      if (stoppedListInHelix == null) {
```
this check no longer required right? Since `getStoppedReplicas()` will not return `null`

jsjtzyy replied:
True, fixed. Thanks.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -136,6 +138,36 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
+      if(stoppedSetUpdated){
+        logger.trace("Updating the stopped list in Helix InstanceConfig");
+        setStoppedResult = setStoppedReplicas(new ArrayList<>(stoppedSet));
+      }else{
```
formatting required

jsjtzyy replied:
done

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -176,33 +236,37 @@ public Task createNewTask(TaskCallbackContext context) {
           "No instance config found for cluster: \"" + clusterName + "\", instance: \"" + instanceName + "\"");
     }
-    return ClusterMapUtils.getSealedReplicas(instanceConfig);
+    logger.trace("Setting InstanceConfig with list of sealed replicas: {}", Arrays.toString(sealedReplicas.toArray()));
```
you don't need the `Arrays.toString(sealedReplicas.toArray())`

jsjtzyy replied:
removed.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -176,33 +236,37 @@ public Task createNewTask(TaskCallbackContext context) {
-    logger.trace("Setting InstanceConfig with list of sealed replicas: {}", Arrays.toString(sealedReplicas.toArray()));
-    instanceConfig.getRecord().setListField(ClusterMapUtils.SEALED_STR, sealedReplicas);
+    logger.trace("Setting InstanceConfig with list of stopped replicas: {}",
+        Arrays.toString(stoppedReplicas.toArray()));
```
same here

jsjtzyy replied:
done.
